### PR TITLE
seed -> cache_seed

### PR DESCRIPTION
**Pull request by sonichi**
## Why are these changes needed?

openai introduces a new param in chat completion api:

seed: This feature is in Beta. If specified, our system will make a best effort to
sample deterministically, such that repeated requests with the same seed and
parameters should return the same result. Determinism is not guaranteed, and you
should refer to the system_fingerprint response parameter to monitor changes
in the backend.
"Determinism is not guaranteed" - that's a reason to keep the seed in autogen, but rename it to cache_seed. Then, we'll be able to support both openai seed and autogen seed.

```[tasklist]
### Tasks
- [x] seed -> cache_seed in code + doc
- [x] add this change to migration guide
- [ ] seed -> cache_seed in notebooks & blogposts for retrieve chat, teachable agent, mathchat
```
The last task won't be performed in this PR. 

## Related issue number

#599

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by qingyun-wu** [comment]
see -> cache_seed in notebooks & tests for llava_agent as well?

**Review by pcdeadeasy — APPROVED** [review]

**Comment by joshkyh** [comment]
I had a look at https://github.com/microsoft/autogen/blob/main/test/agentchat/test_conversable_agent.py
I didn't find a test case for reproducible replies. I think we don't need to test use_cache directly on OpenAI's API (that would be tested by OpenAI), but I think it might be a good idea to add a test to test for reproducible, identical replies on the Autogen class, like conversable agent? 

Perhaps we can test for:
if use_cache is set: then we expect at least 2 out of 5 calls from the conversable agent to be identical, because there is no gurantee.
else: we expect at most 3 out of 5 calls from the agent to be identical

The above numbers are arbitrary and dependent on the prompt/temperature (to be set in the test case). Any thoughts?

File: autogen/oai/completion.py [hunk 1]
```diff
@@ -773,7 +773,7 @@ def yes_or_no_filter(context, config, response):
                 Besides the parameters for the openai API call, it can also contain:
                 - `max_retry_period` (int): the total time (in seconds) allowed for retrying failed requests.
                 - `retry_wait_time` (int): the time interval to wait (in seconds) before retrying a failed request.
-                - `seed` (int) for the cache. This is useful when implementing "controlled randomness" for the completion.
+                - `cache_seed` (int) for the cache. This is useful when implementing "controlled randomness" for the completion.
```
Comment (joshkyh) [inline autogen/oai/completion.py:776]: Perhaps we want to clarify whether cache_seed (int) have an interaction with use_cache (boolean). I think they are independent, right?

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb [hunk 2]
```diff
@@ -359,7 +359,7 @@
     "\n",
     "\n",
     "# Create the manager\n",
-    "llm_config = {\"config_list\": config_list_gpt4, \"seed\": 42, \"use_cache\":False}  # use_cache is False because we want to observe if there is any communication pattern difference if we reran the group chat.\n",
+    "llm_config = {\"config_list\": config_list_gpt4, \"cache_seed\": None}  # cache_seed is None because we want to observe if there is any communication pattern difference if we reran the group chat.\n",
```
Comment (joshkyh) [inline notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb:362]: The answer to whether cache_seed have an interaction with use_cache would determine whether we need to set use_cache to False here.

File: autogen/oai/client.py [hunk 3]
```diff
@@ -219,12 +219,12 @@ def yes_or_no_filter(context, response):
             self._process_for_azure(create_config, extra_kwargs, "extra")
             # construct the create params
             params = self._construct_create_params(create_config, extra_kwargs)
-            # get the seed, filter_func and context
-            seed = extra_kwargs.get("seed", 41)
+            # get the cache_seed, filter_func and context
+            cache_seed = extra_kwargs.get("cache_seed", 41)
```
Comment (joshkyh) [inline autogen/oai/client.py:223]: I wonder if we should default to 41, if use_cache is set to False (not cache_seed), i.e., the user don't want cache/identical replies, shouldn't we change 41 to a random 6-digit number?

**Review by joshkyh — COMMENTED** [review]
Just my two-cents opinion, I might not know the package intricacies well enough. Thanks fo the invite to review.

File: autogen/oai/completion.py (hunk 1 above)
Comment (sonichi) [reply autogen/oai/completion.py:776]: completion.py is only kept for legacy openai package.

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb (hunk 2 above)
Comment (sonichi) [reply notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb:362]: `use_cache` is deprecated in pyautogen v0.2.

File: autogen/oai/client.py (hunk 3 above)
Comment (sonichi) [reply autogen/oai/client.py:223]: `use_cache` is deprecated in pyautogen v0.2.

Comment (joshkyh) [reply autogen/oai/client.py:223]: Yeap, OK, if you think that it should default to 41 when the user sets cache_seed to None, just bringing it to your attention.  :)

I'm not sure whether defaulting to 41 implies local caching.

Comment (sonichi) [reply autogen/oai/client.py:223]: Yes, any integer implies local caching.

**Pull request merged**